core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/util/KeyStoreRegistryPersistenceManager.java

```diff
@@ -0,0 +1,373 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.core.util;
+
+import org.apache.axiom.om.util.UUIDGenerator;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.core.RegistryResources;
+import org.wso2.carbon.core.internal.CarbonCoreDataHolder;
+import org.wso2.carbon.core.security.KeyStoreMetadata;
+import org.wso2.carbon.registry.api.Registry;
+import org.wso2.carbon.registry.api.RegistryException;
+import org.wso2.carbon.registry.api.Resource;
+import org.wso2.carbon.registry.core.Collection;
+import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
+import org.wso2.carbon.utils.security.KeystoreUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * This implementation handles the keystore management in the Registry.
+ */
+public class KeyStoreRegistryPersistenceManager implements KeyStorePersistenceManager {
+
+    private static final Log LOG = LogFactory.getLog(KeyStoreRegistryPersistenceManager.class);
+    private final Registry registry;
+    private static final String REGISTRY_PATH_SEPARATOR = "/";
+
+    private static final String ASSOCIATION_TENANT_KS_PUB_KEY = "assoc.tenant.ks.pub.key";
+    private static final String PROP_TENANT_PUB_KEY_FILE_NAME_APPENDER = "tenant.pub.key.file.name.appender";
+    private final int tenantId;
+
+    public KeyStoreRegistryPersistenceManager(int tenantId) {
+
+        this.registry = getGovernanceRegistry(tenantId);
+        this.tenantId = tenantId;
+    }
+
+    /**
+     * Add the key store to the registry.
+     *
+     * @param keyStore                Key store.
+     * @param keyStoreName            Name of the key store.
+     * @param keystoreContent         Content of the key store.
+     * @param provider                Provider of the key store.
+     * @param encryptedPassword       Encrypted password of the key store.
+     * @param encryptedPrivateKeyPass Encrypted private key password.
+     * @param pvtKeyAlias             Private key alias.
+     * @throws SecurityException If an error occurs while adding the key store.
+     */
+    public void addKeystore(KeyStore keyStore, String keyStoreName, byte[] keystoreContent, String provider,
```

Comment:
Shall we make the methods package-private?

Reply:
The keystore.perisistence pacakge will not be exported.